### ISSUE TITLE
Fatal when using rua_get_level_caps()

### ISSUE THREAD

**Issue by JoryHogeveen** [intoxstudio/restrict-user-access#15]
Hi Joachim,

I get this error when using `rua_get_level_caps()`:
```
Fatal error: Uncaught Error: Class 'WPCAObjectManager' not found in /wp-content/plugins/restrict-user-access/level.php:219

Stack trace:
#0 /wp-content/plugins/restrict-user-access/level.php(206): RUA_Level_Manager->_init_metadata() 
#1 /wp-content/plugins/restrict-user-access/level.php(724): RUA_Level_Manager->metadata() 
#2 /wp-content/plugins/restrict-user-access/api/level.php(138): RUA_Level_Manager->get_levels_caps(Array) 
#3 /wp-content/plugins/view-admin-as/modules/class-restrict-user-access.php(551): rua_get_level_caps(495, true) 
#4 /wp-content/plugins/view-admin-as/modules/class-restrict-user-access.php(155): VAA_View_Admin_As_RUA->get_level_caps(495, true) 
#5 /wp-includes/ in /wp-content/plugins/restrict-user-access/level.php on line 219
```

It has to be a new error since I never encountered it before. It currently happens when I switch to an access level. View Admin As fetches the level caps to set the proper view.

**Comment by intoxstudio** [comment]
Thanks for reporting this!

The wp-content-aware-engine library is loaded pretty early in a page request (plugins_loaded, priority -1). Is there a reason switching levels is done before this?

The problem with calling WPCALoader directly is that it might then load an older version of the library if 2 plugins use it (right now Restrict User Access and Content Aware Sidebars both rely on it).

The "wpca/loaded" action is run immediately after the library has been loaded in "plugins_loaded". Could that hook be used when switching levels?

**Comment by JoryHogeveen** [comment]
Actually, I call `rua_get_level_caps()` after `plugins_loaded` as aswell, only at a lower priority. So I doub't there will be any problem.
I could change it to this but I doubt that will be needed:
```php
if ( ! did_action( 'plugins_loaded' ) {
    return;
}
```

**Comment by intoxstudio** [comment]
Lower than -1? I think a better option would be to just load WPCALoader even before that then, instead of loading it on demand in specific functions. It's actually best to make it load as early as possible (PHP_INT_MIN perhaps?)

**Comment by JoryHogeveen** [comment]
`PHP_INT_MIN` is PHP 7.0+ only so that can't be used in WP.
I don't think there would be a problem in loading it in `_init_metadata()` since this function will only be called once.
Also, `WPCALoader::load();` will return if it's already reloaded.

Why would it load an earlier version if it's loaded before `plugins_loaded` ?

**Comment by intoxstudio** [comment]
I see that now, that's a bummer. I think the problem still is that if another plugin define a newer version of wp-content-aware-engine but its code is executed after Restrict User Access (I think plugins are executed in alphabetical order?), that plugin might not work properly if it relies on a newer version of the library.

Would it solve it to load the library before -99999? It basically just registers a new post type, and sets up modules to handle conditional logic later (in Restrict User Access used for restrictions). It doesn't do anything outside hooks.

**Comment by JoryHogeveen** [comment]
Hmmm, could be a fix, as long as it doesn't do any `current_user_can()` checks etc.
I could also just include the `load()` call in my plugin RUA class and just close this issue, what do you think?

**Comment by JoryHogeveen** [comment]
Added the fix in View Admin As for now since I'm planning to release soon.

**Comment by intoxstudio** [comment]
Sorry for the delay here. No capability checks are made outside any hooks, so it should be safe to load it earlier than View Admin As.

I think that is the only robust solution, as calling `WPCALoader::load()` directly might cause an earlier version to be loaded if 2 plugins rely on it, and then we will run into other problems later.

**Comment by JoryHogeveen** [comment]
Okidoki!

**Comment by intoxstudio** [comment]
Fixed in #d6c09ab, now wp-content-aware-engine is loaded much earlier, priority -999999
